Add accessibility guidelines to the style guide (#14292)
kubernetes/website: content/en/docs/contribute/style/style-guide.md

  <tr><td>The <code>kubectl run</code> command creates a Deployment.</td><td>The "kubectl run" command creates a Deployment.</td></tr>
  <tr><td>For declarative management, use <code>kubectl apply</code>.</td><td>For declarative management, use "kubectl apply".</td></tr>
  <tr><td>Enclose code samples with triple backticks. <code>(```)</code></td><td>Enclose code samples with any other syntax.</td></tr>

>   <tr><td>Use single backticks to enclose inline code. For example, `var example = true`.</td><td>Use two asterisks (**) or an underscore (_) to enclose inline code. For example, **var example = true**.</td></tr><tr><td>Use triple backticks before and after a multi-line block of code for fenced code blocks.</td><td>Use multi-line blocks of code to create diagrams, flowcharts, or other illustrations.</td></tr><tr><td>Use meaningful variable names that have a context.</td><td>Use variable names such as 'foo','bar', and 'baz' that are not meaningful and lack context.</td></tr><tr><td>Remove trailing spaces in the code.</td><td>Add trailing spaces in the code, because the screen reader will readout the spaces as well.</td></tr>

Reviewer: ```suggestion
   <tr><td>Use single backticks to enclose inline code. For example, `var example = true`.</td><td>Use two asterisks (**) or an underscore (_) to enclose inline code. For example, **var example = true**.</td></tr><tr><td>Use triple backticks before and after a multi-line block of code for fenced code blocks.</td><td>Use multi-line blocks of code to create diagrams, flowcharts, or other illustrations.</td></tr><tr><td>Use meaningful variable names that have a context.</td><td>Use variable names such as 'foo','bar', and 'baz' that are not meaningful and lack context.</td></tr><tr><td>Remove trailing spaces in the code.</td><td>Add trailing spaces in the code, where these are important, because the screen reader will read out the spaces as well.</td></tr>
```

(adding _where these are important_ and changing “readout” to “read out”)

>   <tr><td>Use single backticks to enclose inline code. For example, `var example = true`.</td><td>Use two asterisks (**) or an underscore (_) to enclose inline code. For example, **var example = true**.</td></tr><tr><td>Use triple backticks before and after a multi-line block of code for fenced code blocks.</td><td>Use multi-line blocks of code to create diagrams, flowcharts, or other illustrations.</td></tr><tr><td>Use meaningful variable names that have a context.</td><td>Use variable names such as 'foo','bar', and 'baz' that are not meaningful and lack context.</td></tr><tr><td>Remove trailing spaces in the code.</td><td>Add trailing spaces in the code, where these are important, because the screen reader will read out the spaces as well.</td></tr>

Reviewer: This is another place where people reading this _might_ not know what a screen reader is in this context.

As a contributor: after reading this I'm not entirely sure where I would need to leave trailing spaces in place.

</table>

{{< note >}} 

> The website supports syntax highlighting for code samples, but specifying a language is optional. Syntax highlighting in the code block should conform to the [contrast guidelines.](https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&showtechniques=141%2C143#contrast-minimum)

Reviewer 2: - Is the language attribute optional or should authors make their best effort to set this attribute? I would rather have the language attribute set for code blocks.
- Do the k8s docs conform to this guideline? I am not sure. How can we check? Personally, I think the syntax highlighting could be improved, especially for the YAML code blocks.

Author: > * Is the language attribute optional or should authors make their best effort to set this attribute? I would rather have the language attribute set for code blocks.
> * Do the k8s docs conform to this guideline? I am not sure. How can we check? Personally, I think the syntax highlighting could be improved, especially for the YAML code blocks.

This has always been optional, please see https://kubernetes.io/docs/contribute/style/style-guide/#inline-code-formatting. I only added information about the guidelines for color contrast. The link that I provided has all the information and methods that can be used to check. This is a guideline and not a requirement.

 - Uses dashes (-) for unordered lists. 
 - Use the number one (1.) for ordered lists. 
 - Use one space after a list item. 
 - Indent nested lists with four spaces (for example, ⋅⋅⋅⋅-). 
 - List items may consist of multiple paragraphs. Each subsequent paragraph in a list item must be indented by either four spaces or one tab.

### Code

<table>
  <tr><th>Do</th><th>Don't</th></tr>
  <tr><td><ul><li>Use single backticks to enclose inline code. For example, `var example = true`.</li><li>Use triple backticks before and after a multi-line block of code for fenced code blocks.</li><li> Use meaningful variable names that have a context.</li><li>Ensure that there are no trailing spaces in the code, unless it is absolutely neccesary.</li></ul></td><td><ul><li>Use multi-line blocks of code to create diagrams, flowcharts, or other illustrations.</li><li>Use variable names such as 'foo','bar', and 'baz' that are not meaningful and lack context.</li>
  </ul>
</td></tr>
</table>

{{< note >}} Syntax highlighting in the code block should conform to the [contrast guidelines.](https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&showtechniques=141%2C143#contrast-minimum)

### Tables

> The semantic purpose of a data table is to present tabular data. Sighted users can quickly scan the table but a screen reader goes through line by line. Proper markup must be added to help the screen reader make the correct associations that a sighted user would.

Reviewer 2: @Rajakavitha1: With this PR, it seems like the style guide mixes general instructions for authors with accessibility specific instructions. You also don't mention the concept of `accessibility` on the page, unless I missed it. Would it be possible to add a section in the style guide just for accessibility? You may want to provide a link to the accessiblity standards or provide rationale for how the k8s docs are incorporating accessibility. Are these guidelines or requirements?
- In a list of items, is consistency with end punctuation more important than using `a period at the end of each list item`?
- A couple of the table columns/rows are not aligned. See the Links table for an example. 
- nit: `Grouping related items not just looks good visually, but also helps the non-sighted users`. Is `non-sighted` the right term or `visually-impaired users`? I don't know, just wondering.

Author: Thanks @kbhawkey !!!! You are right!!! The idea was to make the current guidelines that we have for writing no different than the accessibility guidelines to write content. We did not want to convey that the accessibility guidelines are good to have or optional. We want to treat these just like the other guidelines that we have.  It might be worth mentioning about writing guidelines to ensure accessibility. I tried to provide the rationale for the guidelines in Line breaks, Headings, Lists, and table. These sections provide guidelines for writing good content in general and address accessibility in specific. The rest of the sections anyways apply in general and address accessibility as well.

In the description, I did provide the link to the accessibility standards that were adopted to Markdown elements (the accessibility level A and AA guidelines only):
https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&currentsidebar=%23col_overview&levels=aaa

-  Yes, you are right in the list items, consistency with end punctuation is important. Not just the period. Could you please help me adopt the guidelines here to our markdown elements (http://web.mit.edu/course/21/21.guide/period.htm). 

- Yes, you are right the tables have just one row with the list of Do and Don't. I did not use separate row for each of the guideline because some of these Do guidelines did not have the corresponding Don't for example: "Ensure that there are no trailing spaces in the code, unless it is absolutely necessary". (Because the screen reader would even read out that space within the code.) Please suggest a better format to present Do and Don't.

- Yes, you are right it is best to avoid the word non-sighted or any other terms that refers to the sight disability. Reworded as suggested.

Reviewer 2: A few more suggestions about front matter and new tables:

Hugo uses [front matter](https://gohugo.io/content-management/front-matter/#readout) to create metadata about a given documentation page. The Kubernetes documentation includes attributes such as [author, title, categories?, weight, and content_template].

```yaml
---
# basic front matter for content, task, tutorial
title:
content_template:
weight:
---
```

```yaml
---
# sample front matter for tutorial
reviewers:
- reviewer1
- reviewer2
title: Runtime Class
content_template: templates/tutorial
weight: 40
---
```
- May want to link to the other style/write-new-topic page containing info about [front matter](https://kubernetes.io/docs/contribute/style/write-new-topic/#adding-the-topic-title-to-the-front-matter), or front matter content to that page? 
- Could we link to a sample from each type of page? There have been a number of blog-related PRs that do not contain the front matter. Or, provide a front matter sample for each type of content page?

- Could the blog content use a separate style guide?

I think Dos and Dont's, as a table, work well when providing real examples -- and there are some mixed in with the new tables. Possibly pull out a few examples and put into a table so that the table looks like the other tables on the page. Do you think a basic list would work for the rest of the content? There is a sample of Dos and Donts from this [accessibility guide](https://developers.google.com/style/accessibility).

Author: @kbhawkey

- I plan to move the fort matter attributes info to https://kubernetes.io/docs/contribute/style/page-templates/

- Do's and Don't list work well. However, to keep it consistent with the rest of the Do and Don't in the table I have rephrased the guidelines.

- I would suggest that we try and keep all the guidelines in one page https://deploy-preview-14292--kubernetes-io-master-staging.netlify.com/docs/contribute/style/style-guide/ only rather than a separate guideline for blog content. We can always integrate the blog content guidelines with our existing guidelines.

Use a single newline to separate block-level content like headings, lists, images, code blocks, and others. The exception is second-level headings, where it should be two newlines. This helps visualize the overall structure of content in a code editor.

### Headings
Remember that screen readers are linear output devices which means they output items on a page one at a time, so if there is a lot of content on your page, applying headings gives the page a structure that allows the user to easily navigate, without having to read loads of content that is of no use, value, or interest to them.
<table>
  <tr><th>Do</th><th>Don't</th></tr>
  <tr><td><ul><li>Use ordered headings to provide a meaningful high-level outline of your content.</li>
  <li>Use only one first-level heading per page.</li>
  <li>For non-blog post content, use pound or hash signs (#), not underlines (--- or ===) to designate first-level headings. </li>
  <li>Use sentence case for headings. For example, **Extend kubectl with plugins** </li></ul></td><td><ul><li>Use first level heading in blog posts, as Hugo will automatically convert the title section of your post’s front matter into a first-level heading.</li>

>   <li>Use headings level 4 through 6, unless it is absolutely necessary. If your content is that detailed, it may need to be broken into separate articles.</li><li>Use title case for headings. For example, **Extend Kubectl With Plugins** </li></ul></td></tr>

Reviewer 3: Hi @Rajakavitha1 :)

+1 for 

> Use sentence case for headings. For example, **Extend kubectl with plugins** ...

Again in line 379 , it is mentioned to use title case for headings. That's a mistake, right?

> Use sentence case for headings. For example, **Extend kubectl with plugins**

Reviewer 3: Ahhhh got it.. sorry about this :)

</td></tr>
</table>

### Lists
Grouping related items not just looks good visually, but also helps the non-sighted users. When a screen reader comes across a list—whether it is an ordered or unordered list—it will be announced to the user that there is a group of list items. The user can then use the arrow keys to cursor up and down between the various items in the list.
Website navigation links can also be marked up as list items; after all they are nothing but a group of related links.

 - Put a period at the end of every list item. 

>  - Uses dashes (-) for unordered lists. 

Reviewer 3: Verifying: Do we use the - or the *? Just checking because I have come across the * so far :)

Author: @shavidissa Thanks, yes * also works :)

{{</* /note */>}}
```

## Markdown elements

### Line breaks

> Use a single newline to separate block-level content like headings, lists, images, code blocks, and others. The exception is second-level headings, where it should be two newlines. This helps visualize the overall structure of content in a code editor.

Reviewer: Should we encourage authors to wrap the text of paragraphs? I know that wrapping makes [GitHub suggested changes](https://help.github.com/en/articles/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) more readable.

Author: Yes, interesting and a very good point. There are pros and cons to wrapping: https://about.gitlab.com/2016/10/11/wrapping-text/. I will leave it as of now. May be later when we decide about wrapping we can update the wrapping guideline.

### Headings

> Remember that screen readers are linear output devices, which means that they output items on a page one at a time. If there is a lot of content on your page, applying headings gives the page a structure that allows the user to easily navigate the page or filter topics of interest.

Reviewer: One more thing. Some people contributing to documentation may not know what “screen reader” means without context.

```suggestion
People accessing this documentation might have trouble seeing, and use a screen reader or other accessibility application. Screen readers in particular are linear output devices: they output items on a page one at a time. If there is a lot of content on a page, you can use headings to give the page an internal structure. A good page structure helps all readers to easily navigate the page or filter topics of interest.
```

  <tr><td>Use pound or hash signs (#) for non-blog post content.</td><td> Use underlines (--- or ===) to designate first-level headings.</td></tr> 
  <tr><td>Use sentence case for headings. For example, <b>Extend kubectl with plugins</b></td><td>Use title case for headings. For example, <b>Extend Kubectl With Plugins</b></td></tr>  
</table>

### Paragraphs

<table>
  <tr><th>Do</th><th>Don't</th></tr>
  <tr><td>Try to keep the paragraphs on the shorter side. 5 to 6 sentences maximum.</td><td>Indent the first paragraph with space characters. For example, ⋅⋅⋅Three spaces before a paragraph will indent it.</td></tr>
  <tr><td>Use three hyphens (---) to create a horizontal rule. Use horizontal rules for breaks in paragraph content. For example, a change of scene in a story, or a shift of topic within a section.</td><td>Use horizontal rules for decoration.</td></tr>
</table>

### Links
 <table>
  <tr><th>Do</th><th>Don't</th></tr>

>   <tr><td>Links should provide context for the content they link to. Certain ports are open on your machines. See <a href="#check-required-ports">Check required ports</a> for more details.</td><td>Use ambiguous terms like “click here”. Certain ports are open on your machines. See <a href="#check-required-ports">here</a> for more details.</td></tr> <tr><td>Use Markdown-style links &lpar;&lsqb;link text&rsqb;&lpar;URL&rpar;&rpar;.</td><td>Use HTML style links &lpar;&lt;link href="/media/examples/link-element-example.css" target="_blank"&gt;Visit our tutorial!&rpar; or create links that open in new tabs or windows. For example, &lsqb;example website&rsqb;&lpar;https://example.com&rpar;{target="_blank"}</td></tr>

Reviewer: The recommended practice and the example seem to be mixed together here(?)

Author: Here's another example: https://deploy-preview-14292--kubernetes-io-master-staging.netlify.com/docs/contribute/style/style-guide/#inline-code-formatting

Just keeping it general under Do and Dont. Because this page is all about guidelines and not requirements.

Reviewer: How about putting the code bits of this lines inside backticks?

>   <tr><td>Links should provide context for the content they link to. For example: Certain ports are open on your machines. See <a href="#check-required-ports">Check required ports</a> for more details.</td><td>Use ambiguous terms like “click here”. For example: Certain ports are open on your machines. See <a href="#check-required-ports">here</a> for more details.</td></tr> <tr><td>Use Markdown-style links &lpar;&lsqb;link text&rsqb;&lpar;URL&rpar;&rpar;.</td><td>Use HTML style links &lpar;&lt;link href="/media/examples/link-element-example.css" target="_blank"&gt;Visit our tutorial!&rpar; or create links that open in new tabs or windows. For example, &lsqb;example website&rsqb;&lpar;https://example.com&rpar;{target="_blank"}</td></tr>

Reviewer: ```suggestion
  <tr><td>Write hyperlinks that give you context for the content they link to. For example: Certain ports are open on your machines. See <a href="#check-required-ports">Check required ports</a> for more details.</td><td>Use ambiguous terms such as “click here”. For example: Certain ports are open on your machines. See <a href="#check-required-ports">here</a> for more details.</td></tr> <tr><td>Write Markdown-style links &lpar;&lsqb;link text&rsqb;&lpar;URL&rpar;&rpar;.</td><td>Write HTML-style links <code>&lpar;&lt;link href="/media/examples/link-element-example.css" target="_blank"&gt;Visit our tutorial!&rpar;</code> or create links that open in new tabs or windows. For example, <code>&lsqb;example website&rsqb;&lpar;https://example.com&rpar;{target="_blank"}</code></td></tr>
```

Also, related: paybe the point about links opening / not opening in the same tab could be on a row by itself?

Author: Yes, but the Do would still be the same right?

</table>

### Lists

> Grouping related items look good visually. When a screen reader comes across a list—whether it is an ordered or unordered list—it will be announced to the user that there is a group of list items. The user can then use the arrow keys to cursor up and down between the various items in the list.

Reviewer: > Grouping related items look good visually.
I'm afraid I'm not sure what this sentence means.

Author: Rephrased the para

Website navigation links can also be marked up as list items; after all they are nothing but a group of related links.

>  - End each item in an list with a period if one or more items in the list are complete sentences. For the sake of parallelism, normally all items or none should be complete sentences.

Reviewer: ```suggestion
 - End each item in a list with a period if one or more items in the list are complete sentences. For the sake of consistency, normally either all items or none should be complete sentences.
```

 
   {{< note >}} If a vertical list completes a sentence begun in the introduction to the list, the final period is omitted unless the items in the list are separated by commas or semicolons.{{< /note >}}
   
 - Use the number one (1.) for ordered lists.
 
 - Use (+), (* ), or (-) for unordered lists.
 
 - Leave a blank line after each list. 
 
 - Indent nested lists with four spaces (for example, ⋅⋅⋅⋅). 
 
 - List items may consist of multiple paragraphs. Each subsequent paragraph in a list item must be indented by either four spaces or one tab.

### Tables

The semantic purpose of a data table is to present tabular data. Sighted users can quickly scan the table but a screen reader goes through line by line. Proper markup must be added to help the screen reader make the correct associations that a sighted user would.

> - Add table captions shortcode. See, [Table captions](/docs/contribute/style/hugo-shortcodes/#table-captions).

Reviewer: ```suggestion
- Add table captions using [Hugo shortcodes](/docs/contribute/style/hugo-shortcodes/#table-captions) for tables.
```

> - Add table captions using [Hugo shortcodes](/docs/contribute/style/hugo-shortcodes/#table-captions) for tables.

Reviewer: ```suggestion
 - Strongly consider adding table captions using [Hugo shortcodes](/docs/contribute/style/hugo-shortcodes/#table-captions) for tables.
```

(because I think there can be rare cases where a caption doesn't really help)

Use a single newline to separate block-level content like headings, lists, images, code blocks, and others. The exception is second-level headings, where it should be two newlines. Second-level headings follow the first-level (or the title) without any preceding paragraphs or texts. A two line spacing helps visualize the overall structure of content in a code editor better.

### Headings

> People accessing this documentation might have trouble seeing, and use a screen reader or other accessibility application. Screen readers in particular are linear output devices: they output items on a page one at a time. If there is a lot of content on a page, you can use headings to give the page an internal structure. A good page structure helps all readers to easily navigate the page or filter topics of interest.

Reviewer 2: How about:
People accessing this documentation may use a screen reader or other assistive technology (AT). [Screen readers](https://en.wikipedia.org/wiki/Screen_reader) are linear output devices.

@Rajakavitha1, if you think linking to the AT or screen reader definition is helfpul, there are probably other links that could be used as well.

  <tr><th>Do</th><th>Don't</th></tr>
  <tr><td>Use only one first-level heading per page.</td><td>Use first level heading in blog posts, as Hugo will automatically convert the title section of your post’s front matter into a first-level heading.</td></tr><tr><td>Use ordered headings to provide a meaningful high-level outline of your content.</td><td>Use headings level 4 through 6, unless it is absolutely necessary. If your content is that detailed, it may need to be broken into separate articles.</td>
  <tr><td>Use pound or hash signs (#) for non-blog post content.</td><td> Use underlines (--- or ===) to designate first-level headings.</td></tr> 

>   <tr><td>Use sentence case for headings. For example, <b>Extend kubectl with plugins</b></td><td>Use title case for headings. For example, <b>Extend Kubectl With Plugins</b></td></tr>  

Reviewer 2: Hello @Rajakavitha1. I read through the content again, in detail, to make sure that I could follow the guidelines. Here are a few areas that may need clarification.

- Line breaks, line 367. I don't fully understand the second sentence about `The exception is second-level headings, where it should be two newlines`
- Headings, line 371, reword or break up the sentences:
Remember that screen readers are linear output devices, which means that they output items on a page one at a time.
If there is a lot of content on your page, applying headings gives the page a structure that allows the user to easily navigate the page[, filtering on topics of interest].
- Headings, line 375, where do we explicitly use `<h1>` or `#`? The hugo template converts the title to the h1 header for all doc pages?
- Paragraphs, line 380, Try to keep paragraphs under 6 sentences.
- Links, line 388, links could show how to code a link, and then show the output. I think this technique is useful. See the example of `Ordered lists` on this page. The example is set up like a cookbook.
- Lists, line 394, `The user can then use the arrow keys to move up and down`. I am not sure about `to cursor`.
- Lists, note, ```{{< note >}} If a vertical list completes a sentence begun in the introduction to the list, the final period is omitted unless the items in the list are separated by commas or semicolons.{{< /note >}}```

This note seems confusing. I think you are saying that it is okay to start a sentence and then add list items, as part of the sentence, where each item ends with a comma or semicolon.

- Tables, line 414, this could be shortened, as the table caption link explains the shortcode.
The semantic purpose of an [html] table is to present tabular data. Sighted users can quickly scan the table but a screen reader goes through line by line[ this is repeated from above]. 

->>What does the table caption do to help the screen reader interpret the output?

[Proper markup must be added to help the screen reader make the correct associations that a sighted user would.]
- Add table caption shortcode. See, [Table captions](/docs/contribute/style/hugo-shortcodes/#table-captions).

Author: Thanks @kbhawkey !!!! As usual, I appreciate the time and efforts you take in reviewing the PR. YOur comments are invaluable. I will look into each of these and commit the changes soon. However, if there is any point that I needs a discussion I will catch up with you on slack.

Author: > * Line breaks, line 367. I don't fully understand the second sentence about `The exception is second-level headings, where it should be two newlines`

H2 usually follows the H1 (or the title) without any preceding paragraphs or texts. A two line spacing helps visualize the overall structure of content in a code editor better.

Author: > his note seems confusing. I think you are saying that it is okay to start a sentence and then add list items, as part of the sentence, where each item ends with a comma or semicolon.

I meant: Ordered lists that are part of an incomplete introductory sentence can be in lowercase and punctuated as if each item was a part of the introductory sentence.

Author: > ->>What does the table caption do to help the screen reader interpret the output?

Imagine reading an entire table cell by cell and then piecing together what the table is trying to tell you.

### Links
 <table>
  <tr><th>Do</th><th>Don't</th></tr>

>   <tr><td>Write hyperlinks that give you context for the content they link to. For example: Certain ports are open on your machines. See <a href="#check-required-ports">Check required ports</a> for more details.</td><td>Use ambiguous terms such as “click here”. For example: Certain ports are open on your machines. See <a href="#check-required-ports">here</a> for more details.</td></tr> <tr><td>Write Markdown-style links &lpar;&lsqb;link text&rsqb;&lpar;URL&rpar;&rpar;. For example, <code>&lsqb;Hugo shortcodes&rsqb;&lpar;/docs/contribute/style/hugo-shortcodes/#table-captions&rpar;</code> and the output is <a href="/docs/contribute/style/hugo-shortcodes/#table-captions">Hugo shortcodes.</td><td>Write HTML-style links <code>&lpar;&lt;link href="/media/examples/link-element-example.css" target="_blank"&gt;Visit our tutorial!&rpar;</code> or create links that open in new tabs or windows. For example, <code>&lsqb;example website&rsqb;&lpar;https://example.com&rpar;{target="_blank"}</code></td></tr>

Reviewer 2: Good

 - End each item in a list with a period if one or more items in the list are complete sentences. For the sake of consistency, normally either all items or none should be complete sentences.
 

>    {{< note >}} Ordered lists that are part of an incomplete introductory sentence can be in lowercase and punctuated as if each item was a part of the introductory sentence.{{< /note >}}

Reviewer 2: Good.
nit: line 395, eachother

### Tables

> The semantic purpose of a data table is to present tabular data. Sighted users can quickly scan the table but a screen reader goes through line by line. Imagine reading an entire table cell by cell and then piecing together what the table is meant for. Hence, it is important to add a proper markup such as a table caption to help the screen reader make the correct associations that a sighted user would. 

Reviewer 2: nit picky: Keep it simple. Do you have any advice on how to write a well written table caption?
Possible edits/ideas:
A table caption is used to create a descriptive title for a data table. Assistive technologies (AT) use the HTML table caption element to identify the table contents to the user within the page structure.
